CK3toEU4/Source/Mappers/ProvinceMapper/ProvinceMapper.cpp

```diff
@@ -0,0 +1,179 @@
+	registerKeys();
+	parseFile("configurables/province_mappings.txt");
+	clearRegisteredKeywords();
+	createMappings();
+	LOG(LogLevel::Info) << "<> " << mappings.size() << " mappings loaded.";
+}
+
+mappers::ProvinceMapper::ProvinceMapper(std::istream& theStream)
+{
+	registerKeys();
+	parseStream(theStream);
+	clearRegisteredKeywords();
+	createMappings();
+}
+
+void mappers::ProvinceMapper::registerKeys()
+{
+	registerRegex(R"([0-9\.]+)", [this](const std::string& unused, std::istream& theStream) {
+		// We support only a single, current version, so eu4-vic2 style multiple versions
+		// have been cut. There should only be a single, 0.0.0.0={} block inside province_mappings.txt
+		if (mappings.empty())
+		{
+			mappings = ProvinceMappingsVersion(theStream).getMappings();
+		}
+		else
+		{
+			commonItems::ignoreItem(unused, theStream);
+		}
+	});
+	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
+}
+
+void mappers::ProvinceMapper::createMappings()
+{
+	for (const auto& mapping: mappings)
+	{
+		// fix deliberate errors where we leave mappings without keys (asian wasteland comes to mind):
+		if (mapping->getCK3Provinces().empty())
+			continue;
+		if (mapping->getEU4Provinces().empty())
+			continue;
+
+		for (auto ck3Number: mapping->getCK3Provinces())
+		{
+			if (ck3Number)
+				CK3ToEU4ProvinceMap.insert(std::make_pair(ck3Number, mapping->getEU4Provinces()));
+		}
+		for (const auto& eu4Number: mapping->getEU4Provinces())
+		{
+			if (eu4Number)
+				EU4ToCK3ProvinceMap.insert(std::make_pair(eu4Number, mapping->getCK3Provinces()));
+		}
+	}
+}
+
+std::map<std::string, std::shared_ptr<CK3::Title>> mappers::ProvinceMapper::getCK3Titles(int eu4ProvinceNumber) const
+{
+	const auto& mapping = EU4ToCK3ProvinceMap.find(eu4ProvinceNumber);
+	std::map<std::string, std::shared_ptr<CK3::Title>> toReturn;
+	if (mapping != EU4ToCK3ProvinceMap.end())
+	{
+		for (auto ck3ID: mapping->second)
+		{
+			if (baroniesToCounties.count(ck3ID))
+			{
+				auto targetTitle = baroniesToCounties.find(ck3ID)->second;
+				toReturn.insert(targetTitle);
+			}
+			else
+			{
+				throw std::runtime_error(
+					 "Province mapper error - requested ck3 province ID " + std::to_string(ck3ID) + " is invalid. We cannot find related county. Farewell.");
+			}
+		}
+		return toReturn;
+	}
+	else
+	{
+		throw std::runtime_error("Province mapper error - requested ck3 province ID does not exist in mappings.");
+	}
+}
+
+std::vector<int> mappers::ProvinceMapper::getEU4ProvinceNumbers(const std::string& ck3CountyName) const
+{
+	const auto& mapping = countiesToBaronies.find(ck3CountyName);
+	if (mapping != countiesToBaronies.end())
+	{
+		const auto& remapping = CK3ToEU4ProvinceMap.find(mapping->second);
+		if (remapping != CK3ToEU4ProvinceMap.end())
+		{
+			return remapping->second;
+		}
+		else
+		{
+			Log(LogLevel::Warning) << "Province Mapper mismatch - " << ck3CountyName << " with baronyID " << mapping->second << " has no mappings!";
```

Comment:
ditto

Reply:
This is not a fail, just a warning. I'm expecting potential mis-lookups from CK3 custom titles and will adjust this to exception or remove warning entirely depending on live data tests.